experimental/packages/opentelemetry-sdk-metrics-base/src/state/MetricStorageRegistry.ts

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { MetricStorage } from './MetricStorage';
+import {
+  getDescriptorIncompatibilityDetails, isDescriptorAsync,
+  isDescriptorCompatibleWith
+} from '../InstrumentDescriptor';
+import * as api from '@opentelemetry/api';
+import { Maybe } from '../utils';
+
+/**
+ * Internal class for storing @{LinkMetricStorage}
+ */
+export class MetricStorageRegistry {
+  private readonly _metricStorageRegistry = new Map<string, MetricStorage>();
+
+  getStorages(): IterableIterator<MetricStorage> {
+    return this._metricStorageRegistry.values();
+  }
+
+  register<T extends MetricStorage>(storage: T): Maybe<T> {
+    const expectedDescriptor = storage.getInstrumentDescriptor();
+
+    // create and register a new one if it does not exist yet.
+    if (!this._metricStorageRegistry.has(expectedDescriptor.name)) {
+      this._metricStorageRegistry.set(expectedDescriptor.name, storage);
+      return storage;
+    }
+
+    // We already checked for existence with has()
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const existingStorage = this._metricStorageRegistry.get(expectedDescriptor.name)!;
+    const existingDescriptor = existingStorage.getInstrumentDescriptor();
+
+    // Return storage if it is compatible.
+    if (isDescriptorCompatibleWith(existingDescriptor, expectedDescriptor)) {
+      // Is compatible, but views for async instruments cannot be registered twice.
+      if(isDescriptorAsync(existingDescriptor)) {
+        api.diag.warn(`A view for an async instrument with the name '${expectedDescriptor.name}' has already been registered.`);
+        return undefined;
+      }
+
+      return (existingStorage as T);
+    }
+
+    // Warn and return undefined if it is incompatible.
```

Comment:
With the change of https://github.com/open-telemetry/opentelemetry-specification/pull/2270, I think we should throw here if the descriptor is not compatible.

> If the registration is semantically not identical, the
  implementation MUST return a registration error to the user when the
  instrument is constructed.

Reply:
@legendecas are you sure we want to throw here? If we do, would it throw during configuration time or when the user calls `createCounter` or whatever method? If the second one, the spec generally tries to discourage throwing when API methods are called in order to avoid crashing a user program.

Reply:
Do you have an exact location of the spec that discourages throwing in API methods? It will be great to comment in spec#2270 before it lands.

Reply:
I think it is this spec [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles). I commented on the line in open-telemetry/opentelemetry-specification#2270 to get some clarification.

Reply:
In https://github.com/open-telemetry/opentelemetry-specification/pull/2317 multiple instrument registrations are allowed even if incompatible. I updated the PR to reflect this, so I'm resolving this conversation for now. :)